src/index.d.ts

```diff
@@ -1,29 +1,112 @@
-export type Version = 0 | 1
-export type Codec = string
-export type Multihash = Buffer
-export type BaseEncodedString = string
-export type MultibaseName = string
-
-export class CID {
-  constructor(version: Version, codec: Codec, multhash: Multihash, multibaseName?: MultibaseName)
-  constructor(cidStr: BaseEncodedString)
+/**
+ * Class representing a CID `<mbase><version><mcodec><mhash>`
+ * , as defined in [ipld/cid](https://github.com/multiformats/cid).
+ */
+declare class CID {
+  /**
+   * Create a new CID.
+   *
+   * The algorithm for argument input is roughly:
+   * ```
+   * if (cid)
+   *   -> create a copy
+   * else if (str)
+   *   if (1st char is on multibase table) -> CID String
+   *   else -> bs58 encoded multihash
+   * else if (Buffer)
+   *   if (1st byte is 0 or 1) -> CID
+   *   else -> multihash
+   * else if (Number)
+   *   -> construct CID by parts
+   * ```
+   *
+   * @example
+   * new CID(<version>, <codec>, <multihash>, <multibaseName>)
+   * new CID(<cidStr>)
+   * new CID(<cid.buffer>)
+   * new CID(<multihash>)
+   * new CID(<bs58 encoded multihash>)
+   * new CID(<cid>)
+   */
+  constructor(version: 0 | 1, codec: string, multhash: Buffer, multibaseName?: string)
+  constructor(cidStr: string)
   constructor(cidBuf: Buffer)
-  constructor(cidMultihash: Multihash)
+  constructor(cidMultihash: Buffer)
   constructor(cid: CID)
-  codec: Codec
-  multihash: Multihash
-  buffer: Buffer
-  prefix: Buffer
+  
+  /**
+   * The codec of the CID.
+   */
+  codec: string
+
+  /**
+   * The multihash of the CID.
+   */
+  multihash: Buffer
+
+  /**
+   * Multibase name as string.
+   */
+  multibaseName: string
+
+  /**
+   * The CID as a `Buffer`
+   */
+  readonly buffer: Buffer
+  
+  /**
+   * The prefix of the CID.
+   */
+  readonly prefix: Buffer
+
+  /**
+   * The version of the CID.
+   */
   version: number
```

Comment:
ditt on on ordering, this is listed first in the code (but honestly this is a very weakly held opinion so feel free to ignore if you have slightly stronger opinions or can't be bothered!)